src/ImageSharp/Processing/Processors/Dithering/IDither.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Processing.Processors.Dithering
+{
+    /// <summary>
+    /// Defines the contract for types that apply dithering to images.
+    /// </summary>
+    public interface IDither
+    {
+        /// <summary>
+        /// Gets the <see cref="Dithering.DitherType"/> which determines whether the
+        /// transformed color should be calculated and supplied to the algorithm.
+        /// </summary>
+        public DitherType DitherType { get; }
+
+        /// <summary>
+        /// Transforms the image applying a dither matrix.
+        /// When <see cref="DitherType"/> is <see cref="DitherType.ErrorDiffusion"/> this
+        /// this method is destructive and will alter the input pixels.
+        /// </summary>
+        /// <param name="image">The image.</param>
+        /// <param name="bounds">The region of interest bounds.</param>
+        /// <param name="source">The source pixel</param>
+        /// <param name="transformed">The transformed pixel</param>
+        /// <param name="x">The column index.</param>
+        /// <param name="y">The row index.</param>
+        /// <param name="bitDepth">The bit depth of the target palette.</param>
+        /// <param name="scale">The dithering scale used to adjust the amount of dither. Range 0..1.</param>
+        /// <typeparam name="TPixel">The pixel format.</typeparam>
+        /// <returns>The dithered result for the source pixel.</returns>
+        TPixel Dither<TPixel>(
+            ImageFrame<TPixel> image,
+            Rectangle bounds,
+            TPixel source,
+            TPixel transformed,
+            int x,
+            int y,
+            int bitDepth,
+            float scale)
+            where TPixel : struct, IPixel<TPixel>;
+    }
```

Comment:
Nice job with the unification, but I'm still wondering if we can find a better abstraction/representation here. The worst thing is that is hard to introduce any optimization with the current design, because of the per-pixel virtual `Dither<T>` method. A it is a clear warning sign for me, that we have 2-element enum defining the two possible algorithms in parallel with the two matching subclasses.

A little bit unorthodox in the world of OO, but I'm wondering about your thoughts on the following idea: Instead of pretending that we are doing something abstract and pluggable in this method, we could simply describe the ditherers by the data (and only the data!) that defines them:

```c#
public struct ErrorDiffusionParameters
{
    public int Offset { get; }

    public DenseMatrix<float> Matrix { get; }
}

public struct OrderedDitherParameters
{
    public DenseMatrix<float> ThresholdMatrix { get; }
}

public sealed class Dither
{
    // Determines which algorithm to use
    public DitherType DitherType { get; }

    // Parameters in case of ErrorDiffusion
    public ErrorDitherParameters? ErrorDiffusionParameters { get; }

    public OrderedDitherParameters? OrderedDitherParameters { get; }

    public static Dither CreateErrorDiffusion(int offset, DenseMatrix<float> matrix) { ... }

    public static Dither CreateOrderedDither(DenseMatrix<float> thresholdMatrix) { .... }
}

public static class KnownDitherings
{
    public static Dither BayerDither2x2 { get; } = CreateBayerDither2x2();

    private static Dither CreateBayerDither2x2()
    {
       DenseMatrix<float> thresholdMatrix;
       
       // calculate thresholdMatrix

       return Dither.CreateOrderedDither(thresholdMatrix);
    }
}

// Implementation could be moved to an internal class.
// Use the per-pixel variants for now, but it would be easy to switch to bulk variants later,
// without getting lost in API design
internal static class DitherExtensions
{
    internal static TPixel ApplyOrderedDither<TPixel>(
            this Dither dither,
            ImageFrame<TPixel> image,
            Rectangle bounds,
            TPixel source,
            TPixel transformed,
            int x,
            int y,
            int bitDepth,
            float scale)
            where TPixel : struct, IPixel<TPixel>
    {
        // ....
    }

    internal static TPixel ApplyErrorDither<TPixel>(
            this Dither dither,
            ImageFrame<TPixel> image,
            Rectangle bounds,
            TPixel source,
            TPixel transformed,
            int x,
            int y,
            int bitDepth,
            float scale)
            where TPixel : struct, IPixel<TPixel>
    {
        // ....
    }
}
```

Reply:
This is very interesting! I'll have a play around later tonight.

Reply:
One issue is the sheer amount of dithering approaches.

- Random,
- BlueNoise
- Halftone

Each one means an extra nullable options property which seems a bit icky.

Reply:
I didn't know that, but it changes the picture a lot, because it means we need real extensibility here. The `DitherType` enum makes it hard to add a new dithering, since it needs touching processor/quantizer code as well.

I think we should keep thinking and find a reasonable abstraction. I have an idea for a double-dispatch trick, I will gist it up later today.

Reply:
Yeah, good point. I'd really like to investigate the possibility of `IDither` implementations being a struct passed via a generic argument also. (_IResampler in the future also in a different PR_).

Been looking at the JIT ASM to see what the difference is between the two approaches and it's chalk and cheese. 

Current class implementation
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdxFAwCyACgCUx0+dcvXZhdigMfAXgYuGAB3Bk4uIKg7IND5bGkAEwAxCAhHBw1qD3Modm47cgz3BgBfGjLqLV1cGHiYBL0w7kiTSqzzHSaImTsASRSIBgAzVIcjN3bsjAALXlwWAYYAkYhM1wqJ107+1IZFowBzGAw1CuLOqy67Xh4GBDHi1xm5hdSWAEEEhLt7tfMKs5tTo3DAyIbYFQMHYQVquEEMT7feG/cqaWi6bDAXAYKAQjCNAYgaHOSYMYoAbRsx2mEASvXEkjsVJmtPpYkkAHkxHwIFx5u8DgdYLhcLwFDBelxJDcbgcHABdc66eGI663e6w7JMADsdwYAGoGOQ/mYAWjOjU6g0yAw4olFiA9rtWuUgA=

Generic struct version
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdxFAwCyACgCUx0+dcvXZhdigMfAXgYuGAB3Bk4uIKgAHnlsaQATADEICAA+OyDQ2ITkiEcHDWoPcyh2bjtyAvcGAF8aOuoaAG1ZAAtvMQAZbGAynn5BACleDABxGEjlOwwATzEYCAAzO2yYJJSHBwBdLV1cGDi1vTDuSKiAFVTq4NaZGAZzhhAGAElcmhMi8x0TiJk7R6LDZGNxfYoYVq8XAsXIMAJAiCFVwNUGuH6PWFGADmMAwaga1R+Vl+dl4PAYCAcn2KZghUJhKRYAEF4vE7JSkeYGgTGrRdGSMDJFtgVK9ctTzAKGCy2VKOfVNHyGD1cBgoCKMMdciA3ilnGCGNUmjZca0IPEXuJJHYTRDzZaxJIAPJiPgQLjQplYrGwXC4XgKGAvLiSMlkrHbQn88ky0nkykS4rEADsFIYAGoGOROWZuYqfvtDvFjqt1hAngxMYqGkA===

Regarding utility methods or extensions. We'll have to build ones that handle the different behaviors in the `FrameQuantizer<TPixel>` and `PaletteDitherProcessor<TPixel>`.

PaletteDitherProcessor
https://github.com/SixLabors/ImageSharp/blob/b9069de3c87b8889918033e92619518c10747e0a/src/ImageSharp/Processing/Processors/Dithering/PaletteDitherProcessor%7BTPixel%7D.cs#L62


FrameQuantizer
https://github.com/SixLabors/ImageSharp/blob/b9069de3c87b8889918033e92619518c10747e0a/src/ImageSharp/Processing/Processors/Quantization/FrameQuantizer%7BTPixel%7D.cs#L176

Reply:
Good idea, I'll keep that in mind!

Reply:
Ok, I mostly figured out `PaletteQuantizer`, but I have no idea how to handle `PaletteDitherProcessor`:
https://gist.github.com/antonfirsov/cbad26bffc39fdd9803b7978a2e87abb

This is getting extremely complex and time-consuming if we want to do it right. Wouldn't it be wiser to close down the API for now, and open a post-RC1 issue?

Reply:
Looks like we're both heading in the same direction. Bear with me, I think I can crack it today.